crates/bonsaidb-local/src/config/argon.rs

```diff
@@ -0,0 +1,165 @@
+    /// Specific argon2 parameters.
+    Params(argon2::ParamsBuilder),
+    /// Automatic configuration based on execution time. This is measured during
+    /// the first `set_password` operation.
+    Timed(TimedArgonParams),
+}
+
+impl ArgonParams {
+    /// Returns the default configuration based on the system information and
+    /// number of hashers. See [`TimedArgonParams`] for more details.
+    pub fn default_for(system: &System, hashers: u32) -> Self {
+        ArgonParams::Timed(TimedArgonParams::default_for(system, hashers))
+    }
+}
+
+/// Automatic configuration based on execution time. This is measured during the
+/// first `set_password`
+#[derive(Debug, Clone)]
+#[must_use]
+pub struct TimedArgonParams {
+    /// The number of lanes (`p`) that the argon algorithm should use.
+    pub lanes: u32,
+    /// The amount of ram each hashing operation should utilize.
+    pub ram_per_hasher: u32,
+    /// The minimum execution time that hashing a password should consume.
+    pub minimum_duration: Duration,
+}
+
+impl Default for TimedArgonParams {
+    /// ## Default Values
+    ///
+    /// When using `TimedArgonParams::default()`, the settings are 4 lanes,
+    /// [`Self::MINIMUM_RAM_PER_HASHER`] of RAM per hasher, and a minimum
+    /// duration of 1 second.
+    ///
+    /// These settings are based on the smallest recommended memory settings
+    /// from [RFC 9106, section 4 "Parameter Choice"][rfc]. The strength of
+    /// Argon2 is derive largely by the amount of RAM dedicated to it, so the
+    /// largest value acceptable should be chosen for `ram_per_hasher`.
+    ///
+    /// [rfc]: https://www.rfc-editor.org/rfc/rfc9106.html#name-parameter-choice
+    fn default() -> Self {
+        Self {
+            lanes: 4,
+            ram_per_hasher: Self::MINIMUM_RAM_PER_HASHER,
+            minimum_duration: Duration::from_secs(1),
+        }
+    }
+}
+
+impl TimedArgonParams {
+    /// The minimum amount of ram to allocate per hasher. This value is
+    /// currently 64MB but will change as the [minimum recommendations][rfc] are
+    /// changed.
+    ///
+    /// [rfc]: https://www.rfc-editor.org/rfc/rfc9106.html#name-parameter-choice
+    pub const MINIMUM_RAM_PER_HASHER: u32 = 64 * 1024 * 1024;
+
+    /// Returns the default configuration based on the system information and
+    /// number of hashers.
+    ///
+    /// The total amount of RAM allocated will be the total system memory
+    /// divided by 16. This allocated amount will be divided equally between the
+    /// hashers. If this number is less than [`Self::MINIMUM_RAM_PER_HASHER`],
+    /// [`Self::MINIMUM_RAM_PER_HASHER`] will be used instead.
+    ///
+    /// For example, if 4 hashers are used on a system with 16GB of RAM, a total
+    /// of 1GB of RAM will be used between 4 hashers, yielding a
+    /// `ram_per_hasher` value of 256MB.
+    ///
+    /// `lanes` defaults to 4.
+    ///
+    /// `minimum_duration` defaults to 1 second.
```

Comment:
The spec only recommends 0.5 seconds btw, though I believe 1 second is more appropriate today.

Reply:
Good point, I'll clarify that's our choice not from the RFC.